<a href="https://colab.research.google.com/github/dmuman/ML-labs/blob/main/lab2/lab2_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Імпорт необхідних для роботи бібліотек

In [313]:
import pandas as pd
import numpy as np

* `pandas` для роботи з даними
* `numpy` для проведемення математичних операцій

# Завдання 1

Необхідно провести оцінку розміру датафрейму та вивести інформацію про кількість пропущених значень у кожному стовпці.

## Зчитування файлу та створення `DataFrame`

In [314]:
# посилання на "сирий" csv файл з гітхабу
url = 'https://raw.githubusercontent.com/OlhaOsypova/Machine-Learning-and-Data-Analytics/refs/heads/main/Topic_3_stat/Task/stack_overflow_2024.csv'

# створення датафрейму
df = pd.read_csv(url)

# виведення перших 5 рядків
df.head()

MainBranch                 Age RemoteWork  \
0  I am a developer by profession  Under 18 years old     Remote   
1  I am a developer by profession     35-44 years old     Remote   
2  I am a developer by profession     45-54 years old     Remote   
3           I am learning to code     18-24 years old        NaN   
4     I code primarily as a hobby  Under 18 years old        NaN   

                                             EdLevel YearsCode YearsCodePro  \
0                          Primary/elementary school       NaN          NaN   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)        20           17   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)        37           27   
3  Some college/university study without earning ...         4          NaN   
4                          Primary/elementary school        10          NaN   

  OrgSize                                            Country  CompTotal  \
0     NaN                           United States of America        NaN   
1     NaN  United Kingdom of Great Britain and Northern I...        NaN   
2     NaN  United Kingdom of Great Britain and Northern I...        NaN   
3     NaN                                             Canada        NaN   
4     NaN                           United States of America        NaN   

              SOVisitFreq  ... TimeAnswering ProfessionalCloud  \
0                     NaN  ...           NaN               NaN   
1  Multiple times per day  ...           NaN               NaN   
2  Multiple times per day  ...           NaN               NaN   
3   Daily or almost daily  ...           NaN               NaN   
4  Multiple times per day  ...           NaN               NaN   

  ProfessionalQuestion JobSatPoints_1 JobSatPoints_4 JobSatPoints_5  \
0                  NaN            NaN            NaN            NaN   
1                  NaN            0.0            0.0            0.0   
2                  NaN            NaN            NaN            NaN   
3                  NaN            NaN            NaN            NaN   
4                  NaN            NaN            NaN            NaN   

            SurveyLength SurveyEase ConvertedCompYearly JobSat  
0                    NaN        NaN                 NaN    NaN  
1                    NaN        NaN                 NaN    NaN  
2  Appropriate in length       Easy                 NaN    NaN  
3               Too long       Easy                 NaN    NaN  
4  Appropriate in length       Easy                 NaN    NaN  

[5 rows x 35 columns]

In [315]:
# виведення останніх 5 рядків
df.tail()

MainBranch              Age  \
58056  I am a developer by profession  18-24 years old   
58057  I am a developer by profession  25-34 years old   
58058  I am a developer by profession  25-34 years old   
58059  I am a developer by profession  18-24 years old   
58060     I code primarily as a hobby  18-24 years old   

                                 RemoteWork  \
58056                                Remote   
58057                                Remote   
58058                             In-person   
58059  Hybrid (some remote, some in-person)   
58060                                   NaN   

                                                 EdLevel YearsCode  \
58056       Bachelor’s degree (B.A., B.S., B.Eng., etc.)         5   
58057                                                NaN       NaN   
58058       Bachelor’s degree (B.A., B.S., B.Eng., etc.)         9   
58059  Secondary school (e.g. American high school, G...         5   
58060                                                NaN       NaN   

      YearsCodePro                   OrgSize  Country  CompTotal  \
58056            3          2 to 9 employees      NaN        NaN   
58057          NaN                       NaN      NaN        NaN   
58058            5  1,000 to 4,999 employees      NaN        NaN   
58059            2        20 to 99 employees  Germany        NaN   
58060          NaN                       NaN      NaN        NaN   

                           SOVisitFreq  ... TimeAnswering ProfessionalCloud  \
58056                              NaN  ...           NaN               NaN   
58057                              NaN  ...           NaN               NaN   
58058                              NaN  ...           NaN               NaN   
58059  A few times per month or weekly  ...           NaN               NaN   
58060            Daily or almost daily  ...           NaN               NaN   

      ProfessionalQuestion JobSatPoints_1 JobSatPoints_4 JobSatPoints_5  \
58056                  NaN            NaN            NaN            NaN   
58057                  NaN            NaN            NaN            NaN   
58058                  NaN            NaN            NaN            NaN   
58059                  NaN            0.0            0.0            0.0   
58060                  NaN            NaN            NaN            NaN   

      SurveyLength SurveyEase ConvertedCompYearly JobSat  
58056          NaN        NaN                 NaN    NaN  
58057          NaN        NaN                 NaN    NaN  
58058          NaN        NaN                 NaN    NaN  
58059          NaN        NaN                 NaN    NaN  
58060          NaN        NaN                 NaN    NaN  

[5 rows x 35 columns]

## Інформація про `DataFrame`

Для отримання інформації буде використано функцію `info()`:

In [316]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58061 entries, 0 to 58060
Data columns (total 35 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   MainBranch            58061 non-null  object 
 1   Age                   58061 non-null  object 
 2   RemoteWork            48673 non-null  object 
 3   EdLevel               53889 non-null  object 
 4   YearsCode             53076 non-null  object 
 5   YearsCodePro          45818 non-null  object 
 6   OrgSize               42136 non-null  object 
 7   Country               52201 non-null  object 
 8   CompTotal             29886 non-null  float64
 9   SOVisitFreq           52758 non-null  object 
 10  SOAccount             52776 non-null  object 
 11  SOPartFreq            40057 non-null  object 
 12  SOComm                52411 non-null  object 
 13  AISelect              53958 non-null  object 
 14  AISent                40721 non-null  object 
 15  AIAcc              

Маємо справу з датафреймом, який має $58061$ рядків (входжень) та $35$ колонок.

Числових колонок - $7$. Категоріальних - $28$.

Майже кожна колонка має в собі пропуски.

## Кількість пропущених даних в кожному стовпці

Для цього буде використано фунції `isnull()` та `sum()`:

In [317]:
df.isnull().sum()

MainBranch                  0
Age                         0
RemoteWork               9388
EdLevel                  4172
YearsCode                4985
YearsCodePro            12243
OrgSize                 15925
Country                  5860
CompTotal               28175
SOVisitFreq              5303
SOAccount                5285
SOPartFreq              18004
SOComm                   5650
AISelect                 4103
AISent                  17340
AIAcc                   24898
AIComplex               25151
AIThreat                18405
TBranch                 18647
ICorPM                  31674
WorkExp                 31787
Knowledge_1             32678
Knowledge_2             33243
Knowledge_3             33186
TimeSearching           32472
TimeAnswering           32531
ProfessionalCloud       32830
ProfessionalQuestion    32555
JobSatPoints_1          32089
JobSatPoints_4          32030
JobSatPoints_5          32009
SurveyLength             8306
SurveyEase               8252
ConvertedCompYearly     37294
JobSat                  32268
dtype: int64

Кожна колонка (крім перших двох) має дуже багато пропусків.

# Завдання 2

Необхідно визначити, які стовпці містять числові дані, але розпізнані як текст.

Потім конвертувати ці стовпці до числового типу даних.

В кінці перевірити результати конвертації, щоб переконатися, що всі дані правильно перетворені.

## Визначення стовпців, які містять числові дані, але розпізнані як текст.

Для цього скористаємось результатами виконання функцій `head()` [тут](https://colab.research.google.com/drive/1vDXM8npB0PULSMRlUYTx5bXZkHIs7Ite#scrollTo=bJx5_393jTiN&line=5&uniqifier=1) та `tail()`[тут](https://colab.research.google.com/drive/1vDXM8npB0PULSMRlUYTx5bXZkHIs7Ite#scrollTo=3X9Qs0xpn2Eq&line=1&uniqifier=1):

В результаті дослідження маємо, що потенційними колонками є `YearsCode` та `YearsCodePro`.

Сюди можна також віднести й колонки `SOAccount`, `AIThreat` та `ICorPM`, де все можна звести до бінарних значень `0` та `1`, але це не стосується завдання напряму.

## Конвертування стовпців до числового типу даних

Для зручності буде відокремлено регулярний вираз (regular expression), який може містити в собі якесь число та будь-який текст з обох боків від нього:

In [318]:
regex = r'[^\d\.]'

А також створено масив (список) необхідних колонок для подальшого ітерування:

In [319]:
columns = [
    'YearsCode',
    'YearsCodePro',
    # 'SOAccount', # колонки, які можна було б бінарно класифікувати
    # 'AIThreat',
    # 'ICorPM'
    ]

In [320]:
# ітерування по колонках
for column in columns:
  # виведення рядка, якщо він має в собі регулярний вираз, ігноруючи NaN
  print(df[column][df[column].str.contains(regex, na=False)])

45       More than 50 years
126      More than 50 years
303      More than 50 years
470        Less than 1 year
555      More than 50 years
                ...        
57741    More than 50 years
57774      Less than 1 year
57787      Less than 1 year
57818      Less than 1 year
58044      Less than 1 year
Name: YearsCode, Length: 731, dtype: object
18       Less than 1 year
37       Less than 1 year
47       Less than 1 year
63       Less than 1 year
145      Less than 1 year
               ...       
57841    Less than 1 year
57878    Less than 1 year
57886    Less than 1 year
57922    Less than 1 year
57970    Less than 1 year
Name: YearsCodePro, Length: 2555, dtype: object


Таким чином було знайдено значення (до того ж однакові в обох випадках), які потрібно замінити на пусті символи.

Для зручності конвертації буде створено словник з необхідними замінами:

In [321]:
# Словник, який заміняє всі непотрібні значення на пусті стрічки
replacing_dict = {
        'More than': '',
        'Less than': '',
        'years': '',
        'year': ''
      }

# тепер буде виконано заміну за допомогою функції replace:
df['YearsCode'] = df['YearsCode'].replace(replacing_dict, regex=True)
df['YearsCodePro'] = df['YearsCodePro'].replace(replacing_dict, regex=True)

# те саме можна виконати і для трьох колонок, про які йшла мова вище. Можна, але непотрібно
# df['SOAccount'] = df['SOAccount'].replace({'Yes': '1', 'No': '0', "0t sure/can't remember": "0"}, regex=True)
# df['AIThreat'] = df['AIThreat'].replace({'Yes': '1', 'No': '0', "I'm not sure": "0"}, regex=True)
# df['ICorPM'] = df['ICorPM'].replace({'Individual contributor': '1', 'People manager': '0'}, regex=True)

Перевірка, чи замінились значення:

In [322]:
for column in columns:
  print(df[column][df[column].str.contains(regex, na=False)])

45        50 
126       50 
303       50 
470        1 
555       50 
         ... 
57741     50 
57774      1 
57787      1 
57818      1 
58044      1 
Name: YearsCode, Length: 731, dtype: object
18        1 
37        1 
47        1 
63        1 
145       1 
        ... 
57841     1 
57878     1 
57886     1 
57922     1 
57970     1 
Name: YearsCodePro, Length: 2555, dtype: object


І справді, лишились тільки числа. От тільки вони в `str` форматі.

Для переведення їх в числовий тип даних, а саме `float`, можна використати метод `astype()`:

In [323]:
# Конвертація даних в числовий формат
for column in columns:
  df[column] = df[column].astype(float)

## Перевірка результатів конвертації

Знову-таки, методом `info()`:

In [324]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58061 entries, 0 to 58060
Data columns (total 35 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   MainBranch            58061 non-null  object 
 1   Age                   58061 non-null  object 
 2   RemoteWork            48673 non-null  object 
 3   EdLevel               53889 non-null  object 
 4   YearsCode             53076 non-null  float64
 5   YearsCodePro          45818 non-null  float64
 6   OrgSize               42136 non-null  object 
 7   Country               52201 non-null  object 
 8   CompTotal             29886 non-null  float64
 9   SOVisitFreq           52758 non-null  object 
 10  SOAccount             52776 non-null  object 
 11  SOPartFreq            40057 non-null  object 
 12  SOComm                52411 non-null  object 
 13  AISelect              53958 non-null  object 
 14  AISent                40721 non-null  object 
 15  AIAcc              

Можна помітити, що дві обраних колонки змінили тип даних на `float64`. Конвертація пройшла успішно.

# Завдання 3

Необхідно провести видалення викидів у числових даних:
* Для кожної числової ознаки визначити нижній і верхній квантилі на рівні $1\%$ та $99\%$.
* Видалити значення, що знаходяться поза цими квантилями (тобто значення, які можна вважати викидами).
* Відобразити кількість видалених рядків для кожної ознаки та зробити висновки про поширеність викидів у даних.

Для початку буде створено масив чисельних колонок. Зробити це можна двома способами:
* масивом `numpy`
* або просто за допомогою функції `select_dtypes()`

У завданні буде використано другий варіант:

In [325]:
# numeric_columns = np.array([column for column in df.columns if df[column].dtype == 'float64'], dtype='object')
numeric_columns = df.select_dtypes(include='float64').columns

# ітерація по колонкаї
for column in numeric_columns:
  print(f'# {column}')
  # встановлення нижнього та верхнього квантилів
  lower_bound = df[column].quantile(0.01)
  upper_bound = df[column].quantile(0.99)

  # виведення кількості рядків до видалення викидів
  print(f'Кількість рядків до видалення викидів {column} : {df.shape[0]}')

  # створення очищеного фрейму, який міститиме дані з колонок
  # тільки, якщо вони більші за нижній квантиль ТА нижчі за верхній
  # АБО має значення NaN
  df_cleaned = df[((df[column] >= lower_bound) & (df[column] <= upper_bound)) | (df[column].isna())]

  # виведення кількості рядків після видалення викидів
  # та кількість видалених рядків
  print(f'Кількість рядків після видалення викидів {column} : {df_cleaned.shape[0]}')
  print(f'Видалено рядків з {column} : {df.shape[0] - df_cleaned.shape[0]}')
  print()

# YearsCode
Кількість рядків до видалення викидів YearsCode : 58061
Кількість рядків після видалення викидів YearsCode : 57536
Видалено рядків з YearsCode : 525

# YearsCodePro
Кількість рядків до видалення викидів YearsCodePro : 58061
Кількість рядків після видалення викидів YearsCodePro : 57735
Видалено рядків з YearsCodePro : 326

# CompTotal
Кількість рядків до видалення викидів CompTotal : 58061
Кількість рядків після видалення викидів CompTotal : 57485
Видалено рядків з CompTotal : 576

# WorkExp
Кількість рядків до видалення викидів WorkExp : 58061
Кількість рядків після видалення викидів WorkExp : 57644
Видалено рядків з WorkExp : 417

# JobSatPoints_1
Кількість рядків до видалення викидів JobSatPoints_1 : 58061
Кількість рядків після видалення викидів JobSatPoints_1 : 58061
Видалено рядків з JobSatPoints_1 : 0

# JobSatPoints_4
Кількість рядків до видалення викидів JobSatPoints_4 : 58061
Кількість рядків після видалення викидів JobSatPoints_4 : 58061
Видалено рядків з JobSatPo

Можна помітити, що було видалено не так вже й багато рядків (що означає, що викидів було небагато). Це число буде вищим, якщо, наприклад, зменшити верхній квантиль до $95\%$, а нижній підняти до $5\%$.

Замінимо дані у фреймі на дані з очищеного фрейму.

Використаємо копію, аби потім не було негараздів:

In [326]:
df = df_cleaned.copy()

# Завдання 4

Необхідно видалити усі стовпці, у яких пропущено $50\%$ або більше значень.

Зробити це можна за допомогою функції `dropna()`, встановивши поріг (`thresh`) на половину довжини колонки та значеня осі (`axis`) на `1`, адже видаляти треба колонки (`0` - це рядки):

In [327]:
df = df.dropna(thresh = len(df) * 0.5, axis=1)
df.info() # виведемо інформацію для порівняння

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58061 entries, 0 to 58060
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   MainBranch    58061 non-null  object 
 1   Age           58061 non-null  object 
 2   RemoteWork    48673 non-null  object 
 3   EdLevel       53889 non-null  object 
 4   YearsCode     53076 non-null  float64
 5   YearsCodePro  45818 non-null  float64
 6   OrgSize       42136 non-null  object 
 7   Country       52201 non-null  object 
 8   CompTotal     29886 non-null  float64
 9   SOVisitFreq   52758 non-null  object 
 10  SOAccount     52776 non-null  object 
 11  SOPartFreq    40057 non-null  object 
 12  SOComm        52411 non-null  object 
 13  AISelect      53958 non-null  object 
 14  AISent        40721 non-null  object 
 15  AIAcc         33163 non-null  object 
 16  AIComplex     32910 non-null  object 
 17  AIThreat      39656 non-null  object 
 18  TBranch       39414 non-nu

Можна помітити, що суттєво скоротилась кількість як числовних, так і категоріальних колонок: $9$ => $3$ та $26$ => $18$ відповідно.

# Завдання 5

Для решти стовпців з пропусками треба заповнити або видалити пропущені значення. Вибір методу заповнення (або видалення) потрібно обґрунтувати для кожного стовпця.

Для початку перевіримо кількість пропусків в кожній колонці:

In [328]:
df.isnull().sum()

MainBranch          0
Age                 0
RemoteWork       9388
EdLevel          4172
YearsCode        4985
YearsCodePro    12243
OrgSize         15925
Country          5860
CompTotal       28175
SOVisitFreq      5303
SOAccount        5285
SOPartFreq      18004
SOComm           5650
AISelect         4103
AISent          17340
AIAcc           24898
AIComplex       25151
AIThreat        18405
TBranch         18647
SurveyLength     8306
SurveyEase       8252
dtype: int64

Можна помітити, що кількість пропусків сягає $10-15$ відсотків. Враховуючи це, пропущені дані можна не видаляти, а просто заповнити іншими значеннями (як от модою, медіаною або середнім значенням).

## Заповнення числових даних

Для початку буде обрано всі числові колонки:

In [329]:
numeric_columns = df.select_dtypes(float).columns
numeric_columns

Index(['YearsCode', 'YearsCodePro', 'CompTotal'], dtype='object')

Всі дані спочатку буде згруповано за віком (колонка `'Age'`) та трансформовано до середнього значення (`mean`) по колонці. Далі пропущені дані буде заповнено цими значеннями.

In [330]:
for column in numeric_columns:
  df.loc[:, column] = df[column].fillna(df.groupby('Age')[column].transform('mean'))

Чому середнє значення? Це досить розповсюджена практика заповнення числових значень. Вона дозволяє в подальшому більш точно розрахувати необхідні значення.

`loc` було використано для безпечного індексування фрейму.

Перевіримо пропущені значення тепер:

In [331]:
df.isnull().sum()

MainBranch          0
Age                 0
RemoteWork       9388
EdLevel          4172
YearsCode           0
YearsCodePro        0
OrgSize         15925
Country          5860
CompTotal           0
SOVisitFreq      5303
SOAccount        5285
SOPartFreq      18004
SOComm           5650
AISelect         4103
AISent          17340
AIAcc           24898
AIComplex       25151
AIThreat        18405
TBranch         18647
SurveyLength     8306
SurveyEase       8252
dtype: int64

Можна помітити, що числові колонки тепер не мають пропущених значень. Тобто, вони усі заповнені.

## Заповнення та видалення текстових даних

Для початку буде просто видено рядки, де більше половини значень - це `NaN`.

Все так само, як і з колонками, тільки параметр `axis` тепер можна не встановлювати (він `0` за замовчуванням):

In [332]:
df = df.dropna(thresh = len(df.columns) * 0.5)

In [333]:
df.isnull().sum() # інформація про пропущені дані

MainBranch          0
Age                 0
RemoteWork       8261
EdLevel           812
YearsCode           0
YearsCodePro        0
OrgSize         11873
Country          1402
CompTotal           0
SOVisitFreq      1254
SOAccount        1232
SOPartFreq      13575
SOComm           1391
AISelect          399
AISent          13367
AIAcc           20663
AIComplex       20878
AIThreat        13787
TBranch         13814
SurveyLength     3505
SurveyEase       3451
dtype: int64

In [334]:
df.info() # загальна інформація по фрейму

<class 'pandas.core.frame.DataFrame'>
Index: 53226 entries, 0 to 58059
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   MainBranch    53226 non-null  object 
 1   Age           53226 non-null  object 
 2   RemoteWork    44965 non-null  object 
 3   EdLevel       52414 non-null  object 
 4   YearsCode     53226 non-null  float64
 5   YearsCodePro  53226 non-null  float64
 6   OrgSize       41353 non-null  object 
 7   Country       51824 non-null  object 
 8   CompTotal     53226 non-null  float64
 9   SOVisitFreq   51972 non-null  object 
 10  SOAccount     51994 non-null  object 
 11  SOPartFreq    39651 non-null  object 
 12  SOComm        51835 non-null  object 
 13  AISelect      52827 non-null  object 
 14  AISent        39859 non-null  object 
 15  AIAcc         32563 non-null  object 
 16  AIComplex     32348 non-null  object 
 17  AIThreat      39439 non-null  object 
 18  TBranch       39412 non-null  o

Можна помітити, що зменшилась загальна кількість рядків: з $58061$ до $53226$ (через це стало менше й пропущених значень).

Тепер оберемо всі категоріальні колонки:

In [335]:
categorical_columns = df.select_dtypes(object).columns
categorical_columns

Index(['MainBranch', 'Age', 'RemoteWork', 'EdLevel', 'OrgSize', 'Country',
       'SOVisitFreq', 'SOAccount', 'SOPartFreq', 'SOComm', 'AISelect',
       'AISent', 'AIAcc', 'AIComplex', 'AIThreat', 'TBranch', 'SurveyLength',
       'SurveyEase'],
      dtype='object')

Відсутні значення для колонки `EdLevel` (рівень освіти) краще буде замінити на `'Not given'` (не надано), адже інакше можуть сформуватись хибні враження щодо рівня освіти респондентів:

* `loc` слугує тим самим

In [336]:
df.loc[:, 'EdLevel'] = df['EdLevel'].fillna('Not given')

Всі інші колонки можна заповнити даними, які зустрічаються найчастіше - модою (`mode`):

In [337]:
# ітерування по колонках
for column in categorical_columns:
  df.loc[:, column] = df[column].fillna(df[column].mode()[0]) # індекс 0, аби отримати тільки значення

In [338]:
df.isnull().sum() # к-сть пропущених значень

MainBranch      0
Age             0
RemoteWork      0
EdLevel         0
YearsCode       0
YearsCodePro    0
OrgSize         0
Country         0
CompTotal       0
SOVisitFreq     0
SOAccount       0
SOPartFreq      0
SOComm          0
AISelect        0
AISent          0
AIAcc           0
AIComplex       0
AIThreat        0
TBranch         0
SurveyLength    0
SurveyEase      0
dtype: int64

Після заміни усіх текстових даних на найбільш популярні значення маємо $0$ пропусків.

# Завдання 6

Треба провести дослідження думок українських IT-спеціалістів щодо штучного інтелекту:
* Створити новий датафрейм, який включатиме тільки респондентів з України.
* Визначити, як українські спеціалісти ставляться до штучного інтелекту:
 * наскільки часто вони використовують AI,
 * як оцінюють його складність,
 * довіру до нього,
 * та чи бачать загрозу у своїй роботі.

Для цього спочатку буде створено фільтр (булеанову маску), який обере тільки ті рядкі, де в колонці `'Country'` вказано `'Ukraine'`.

Далі буде створено масив необхідних колонок.

І після цього буде створено новий фрейм, який міститиме лише необхідні дані:

In [339]:
df_ukraine_filter = df[df['Country'] == 'Ukraine']
AI_columns = ['AISelect', 'AISent', 'AIAcc', 'AIComplex', 'AIThreat']

df_ukraine = df_ukraine_filter[AI_columns]

Проведемо дослідження цього фрейму.

Спочатку просто перевіривши форму (shape), аби дізнатись кількість респондентів з України, а потім за допомогою фукнції `describe()`:

In [340]:
df_ukraine.shape

(2373, 5)

Отже, з України у опитуванні взяли участь 2373 особи. Принаймні, це кількість людей у вже очищеному фреймі.

In [341]:
df_ukraine.describe(include=[object])

AISelect     AISent           AIAcc  \
count      2373       2373            2373   
unique        3          6               5   
top         Yes  Favorable  Somewhat trust   
freq       1716       1287            1375   

                                            AIComplex AIThreat  
count                                            2373     2373  
unique                                              5        3  
top     Good, but not great at handling complex tasks       No  
freq                                             1196     1595

Щодо аналізу самих відповідей:
* більша половина респондентів використовує штучний інтелект в роботі ($1716$ людей з $2373$),
* також більша половина ($1287$ з $2373$) позитивно ставиться до штучного інтелекту,
* $1375$ з $2373$ людей дещо довіряють ШІ,
* $1196$ респондентів вважають, що ШІ справляється з задачами, але він не дуже добрий для комплексних задач,
* $1595$ респондентів не вважають ШІ загрозою для себе.

Тепер колонку `'County'` можна просто видалити, адже вона має забагато категорій:

In [342]:
df = df.drop('Country', axis=1)

In [343]:
df.head() # перевірка видалення

MainBranch                 Age  \
0  I am a developer by profession  Under 18 years old   
1  I am a developer by profession     35-44 years old   
2  I am a developer by profession     45-54 years old   
3           I am learning to code     18-24 years old   
4     I code primarily as a hobby  Under 18 years old   

                             RemoteWork  \
0                                Remote   
1                                Remote   
2                                Remote   
3  Hybrid (some remote, some in-person)   
4  Hybrid (some remote, some in-person)   

                                             EdLevel  YearsCode  YearsCodePro  \
0                          Primary/elementary school   4.473151      2.038123   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  20.000000     17.000000   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  37.000000     27.000000   
3  Some college/university study without earning ...   4.000000      2.292955   
4                          Primary/elementary school  10.000000      2.038123   

              OrgSize     CompTotal             SOVisitFreq SOAccount  \
0  20 to 99 employees  2.723611e+07    A few times per week       Yes   
1  20 to 99 employees  1.208021e+61  Multiple times per day       Yes   
2  20 to 99 employees  5.529107e+06  Multiple times per day       Yes   
3  20 to 99 employees  1.963190e+24   Daily or almost daily        No   
4  20 to 99 employees  2.723611e+07  Multiple times per day       Yes   

                            SOPartFreq           SOComm  \
0  Less than once per month or monthly   No, not really   
1               Multiple times per day  Yes, definitely   
2               Multiple times per day  Yes, definitely   
3  Less than once per month or monthly   No, not really   
4               Multiple times per day  Yes, definitely   

                  AISelect          AISent           AIAcc  \
0                      Yes  Very favorable  Somewhat trust   
1  No, and I don't plan to       Favorable  Somewhat trust   
2  No, and I don't plan to       Favorable  Somewhat trust   
3                      Yes  Very favorable  Somewhat trust   
4                      Yes       Favorable  Somewhat trust   

                                       AIComplex AIThreat TBranch  \
0  Good, but not great at handling complex tasks       No      No   
1  Good, but not great at handling complex tasks       No     Yes   
2  Good, but not great at handling complex tasks       No      No   
3                  Bad at handling complex tasks       No     Yes   
4  Good, but not great at handling complex tasks       No     Yes   

            SurveyLength SurveyEase  
0  Appropriate in length       Easy  
1  Appropriate in length       Easy  
2  Appropriate in length       Easy  
3               Too long       Easy  
4  Appropriate in length       Easy

# Завдання 7

Необхідно перетворити усі категоріальні змінні за допомогою методу One-Hot Encoding.

Для цього спочатку буде обрано всі категоріальні та числові колонки:

In [344]:
categorical_columns = df.select_dtypes(object).columns
numeric_columns = df.select_dtypes(include=[float, int]).columns

Тепер перетворимо категоріальні змінні. Для цього буде імпортовано `OneHotEncoding` з модуля `preprocessing` бібліотеки `sklearn`:

In [345]:
# імпорт та ініціалізація енкодера
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(sparse_output=False, drop='first')
# sparse_output=False: Встановлює повернення щільної матриці (звичайний 2D масив), а не розрідженої (спрощує обробку).
# drop='first': Видаляє перший стовпець для кожної категоріальної змінної, щоб уникнути дублювання інформації.

In [346]:
# застосування OneHotEncoder до категоріальних стовпців
encoded_columns = encoder.fit_transform(df[categorical_columns])

In [347]:
# створення нового фрейму з перекодованими змінними та назвами колонок
encoded_df = pd.DataFrame(encoded_columns, columns=encoder.get_feature_names_out(categorical_columns))

In [348]:
# об'єднання початкового фрейму з перекодованими стовпцями
df = df.drop(columns=categorical_columns)
df = pd.concat([df, encoded_df], axis=1)

In [349]:
# перевірка результатів
df.head()

YearsCode  YearsCodePro     CompTotal  MainBranch_I am learning to code  \
0   4.473151      2.038123  2.723611e+07                               0.0   
1  20.000000     17.000000  1.208021e+61                               0.0   
2  37.000000     27.000000  5.529107e+06                               0.0   
3   4.000000      2.292955  1.963190e+24                               1.0   
4  10.000000      2.038123  2.723611e+07                               0.0   

   MainBranch_I am not primarily a developer, but I write code sometimes as part of my work/studies  \
0                                                0.0                                                  
1                                                0.0                                                  
2                                                0.0                                                  
3                                                0.0                                                  
4                                                0.0                                                  

   MainBranch_I code primarily as a hobby  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     1.0   

   MainBranch_I used to be a developer by profession, but no longer am  \
0                                                0.0                     
1                                                0.0                     
2                                                0.0                     
3                                                0.0                     
4                                                0.0                     

   Age_25-34 years old  Age_35-44 years old  Age_45-54 years old  ...  \
0                  0.0                  0.0                  0.0  ...   
1                  0.0                  1.0                  0.0  ...   
2                  0.0                  0.0                  1.0  ...   
3                  0.0                  0.0                  0.0  ...   
4                  0.0                  0.0                  0.0  ...   

   AIComplex_Neither good or bad at handling complex tasks  \
0                                                0.0         
1                                                0.0         
2                                                0.0         
3                                                0.0         
4                                                0.0         

   AIComplex_Very poor at handling complex tasks  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   

   AIComplex_Very well at handling complex tasks  AIThreat_No  AIThreat_Yes  \
0                                            0.0          1.0           0.0   
1                                            0.0          1.0           0.0   
2                                            0.0          1.0           0.0   
3                                            0.0          1.0           0.0   
4                                            0.0          1.0           0.0   

   TBranch_Yes  SurveyLength_Too long  SurveyLength_Too short  \
0          0.0                    0.0                     0.0   
1          1.0                    0.0                     0.0   
2          0.0                    0.0                     0.0   
3          1.0                    1.0                     0.0   
4          1.0                    0.0                     0.0   

   SurveyEase_Easy  SurveyEase_Neither easy nor difficult  
0              1.0                                    0.0  
1              1.0                                    0.0  
2              1.0          

# Завдання 8

Проведення стандартизації та нормалізації.

Для стандартизації буде використано `StandardScaler` з того ж `sklearn.preprocessing`.
Для нормалізації - `MinMaxScaler`.

## `StandardScaler`

Спочатку проведемо стандартизацію:

In [350]:
# імпорт та ініціалізація скейлера
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [351]:
# застосування StandardScaler до кількісних стовпців
standardized_columns = scaler.fit_transform(df[numeric_columns])

In [352]:
# перетворення стандартизованих даних в датафрейм з відповідними назвами стовпців
standartized_df = pd.DataFrame(standardized_columns, columns=numeric_columns, index=df.index)

In [353]:
# видалення початкових кількісних ознак
df = df.drop(columns=numeric_columns)

In [354]:
# додавання стандартизованих ознак назад в датафрейм
df = pd.concat([df, standartized_df], axis=1)

In [355]:
# перевірка результатів
df.head()

MainBranch_I am learning to code  \
0                               0.0   
1                               0.0   
2                               0.0   
3                               1.0   
4                               0.0   

   MainBranch_I am not primarily a developer, but I write code sometimes as part of my work/studies  \
0                                                0.0                                                  
1                                                0.0                                                  
2                                                0.0                                                  
3                                                0.0                                                  
4                                                0.0                                                  

   MainBranch_I code primarily as a hobby  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     1.0   

   MainBranch_I used to be a developer by profession, but no longer am  \
0                                                0.0                     
1                                                0.0                     
2                                                0.0                     
3                                                0.0                     
4                                                0.0                     

   Age_25-34 years old  Age_35-44 years old  Age_45-54 years old  \
0                  0.0                  0.0                  0.0   
1                  0.0                  1.0                  0.0   
2                  0.0                  0.0                  1.0   
3                  0.0                  0.0                  0.0   
4                  0.0                  0.0                  0.0   

   Age_55-64 years old  Age_65 years or older  Age_Prefer not to say  ...  \
0                  0.0                    0.0                    0.0  ...   
1                  0.0                    0.0                    0.0  ...   
2                  0.0                    0.0                    0.0  ...   
3                  0.0                    0.0                    0.0  ...   
4                  0.0                    0.0                    0.0  ...   

   AIThreat_No  AIThreat_Yes  TBranch_Yes  SurveyLength_Too long  \
0          1.0           0.0          0.0                    0.0   
1          1.0           0.0          1.0                    0.0   
2          1.0           0.0          0.0                    0.0   
3          1.0           0.0          1.0                    1.0   
4          1.0           0.0          1.0                    0.0   

   SurveyLength_Too short  SurveyEase_Easy  \
0                     0.0              1.0   
1                     0.0              1.0   
2                     0.0              1.0   
3                     0.0              1.0   
4                     0.0              1.0   

   SurveyEase_Neither easy nor difficult  YearsCode  YearsCodePro  CompTotal  
0                                    0.0  -0.923063     -0.865931  -0.006507  
1                                    0.0   0.540758      0.776674  -0.006507  
2                                    0.0   2.143463      1.874535  -0.006507  
3                                    0.0  -0.967670     -0.837954  -0.006507  
4                                    0.0  -0.402009     -0.865931  -0.006507  

[5 rows x 71 columns]

## `MinMaxScaler`

Тепер нормалізація:

In [356]:
# імпорт та ініціалізація скейлера
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

In [357]:
# застосування MinMaxScaler до кількісних стовпців
normalized_columns = scaler.fit_transform(df[numeric_columns])

In [358]:
# перетворення нормалізованих даних в датафрейм з відповідними назвами стовпців
normalized_df = pd.DataFrame(normalized_columns, columns=numeric_columns, index=df.index)

In [359]:
# видалення початкових кількісних ознак
df = df.drop(columns=numeric_columns)

In [360]:
# додавання нормалізованих ознак назад в датафрейм
df = pd.concat([df, normalized_df], axis=1)

In [361]:
# перевірка результатів
df.head()

MainBranch_I am learning to code  \
0                               0.0   
1                               0.0   
2                               0.0   
3                               1.0   
4                               0.0   

   MainBranch_I am not primarily a developer, but I write code sometimes as part of my work/studies  \
0                                                0.0                                                  
1                                                0.0                                                  
2                                                0.0                                                  
3                                                0.0                                                  
4                                                0.0                                                  

   MainBranch_I code primarily as a hobby  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     1.0   

   MainBranch_I used to be a developer by profession, but no longer am  \
0                                                0.0                     
1                                                0.0                     
2                                                0.0                     
3                                                0.0                     
4                                                0.0                     

   Age_25-34 years old  Age_35-44 years old  Age_45-54 years old  \
0                  0.0                  0.0                  0.0   
1                  0.0                  1.0                  0.0   
2                  0.0                  0.0                  1.0   
3                  0.0                  0.0                  0.0   
4                  0.0                  0.0                  0.0   

   Age_55-64 years old  Age_65 years or older  Age_Prefer not to say  ...  \
0                  0.0                    0.0                    0.0  ...   
1                  0.0                    0.0                    0.0  ...   
2                  0.0                    0.0                    0.0  ...   
3                  0.0                    0.0                    0.0  ...   
4                  0.0                    0.0                    0.0  ...   

   AIThreat_No  AIThreat_Yes  TBranch_Yes  SurveyLength_Too long  \
0          1.0           0.0          0.0                    0.0   
1          1.0           0.0          1.0                    0.0   
2          1.0           0.0          0.0                    0.0   
3          1.0           0.0          1.0                    1.0   
4          1.0           0.0          1.0                    0.0   

   SurveyLength_Too short  SurveyEase_Easy  \
0                     0.0              1.0   
1                     0.0              1.0   
2                     0.0              1.0   
3                     0.0              1.0   
4                     0.0              1.0   

   SurveyEase_Neither easy nor difficult  YearsCode  YearsCodePro  CompTotal  
0                                    0.0   0.070881      0.021186        0.0  
1                                    0.0   0.387755      0.326531        0.0  
2                                    0.0   0.734694      0.530612        0.0  
3                                    0.0   0.061224      0.026387        0.0  
4                                    0.0   0.183673      0.021186        0.0  

[5 rows x 71 columns]

# Висновок

Під час виконання даної лабораторної роботи я навчився правильно очищати датафрейм від пропущених даних. Встановлювати верхній та нижній пороги, замінювати пропущені значення різними способами. А також навчився перетворювати категоріальні змінні та проводити нормалізацію й стандартизацію кількісних даних. Відкрив для себе три нових методи `sklearn`: `OneHotEncoding`, `MinMaxScaler` та `StandardScaler`.

А також виконав усі необхідні завдання.